test/unit/Cardano/Wallet/CoinSelection/LargestFirstSpec.hs

```diff
@@ -0,0 +1,336 @@
+    resultOne <- runExceptT $ largestFirst
+                 (defaultCoinSelectionOptions 100)
+                 utxo
+                 txOuts
+    resultTwo <- runExceptT $ largestFirst
+                 (defaultCoinSelectionOptions 100)
+                 utxo
+                 txOuts
+
+    resultOne `shouldBe` resultTwo
+
+propAtLeast
+    :: CoveringCase
+    -> Property
+propAtLeast (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+
+    L.length
+        (concatMap (\ (CoinSelection inps _ _) -> inps) $ rights [result])
+        `shouldSatisfy`
+        (>= NE.length txOuts)
+
+propInputDecreasingOrder
+    :: CoveringCase
+    -> Property
+propInputDecreasingOrder (CoveringCase (utxo, txOuts)) =
+    (condCoinsCovering txOuts utxo) ==> monadicIO $ liftIO $ do
+    result <- runExceptT $ largestFirst
+              (defaultCoinSelectionOptions 100)
+              utxo
+              txOuts
+    let utxoEntriesInCoinSelection =
+            concatMap (\ (CoinSelection inps _ _) -> inps)
+            $ rights [result]
+    let utxo' =
+            Map.toList (getUTxO utxo) L.\\ utxoEntriesInCoinSelection
+    let getExtremumValue f = f . map (getCoin . coin . snd)
+
+    if L.null utxo' then utxo' `shouldBe` utxo' else
+        (getExtremumValue L.minimum utxoEntriesInCoinSelection)
+        `shouldSatisfy` (>= (getExtremumValue L.maximum utxo'))
+
+{-------------------------------------------------------------------------------
+                                  Test Data
+-------------------------------------------------------------------------------}
+
+defaultCoinSelectionOptions
+    :: Word64
+    -> CoinSelectionOptions
+defaultCoinSelectionOptions n = CoinSelectionOptions
+    { estimateFee = \_ _ -> Coin 0
+    , dustThreshold = Coin 0
+    , maximumNumberOfInputs = n
+    }
+
+
+-- Check if there are enough covering utxo entries
+-- in the utxo to cover all transaction outputs
+condCoinsCovering
+    :: NonEmpty TxOut
+    -> UTxO
+    -> Bool
+condCoinsCovering txOuts utxo =
+    let
+        len = NE.length txOuts
+    in L.length (utxoCovering txOuts utxo) >= len
+
+-- Select all UTxO entries that could cover all
+-- transaction outputs sorted from the smallest one
+utxoCovering
+    :: NonEmpty TxOut
+    -> UTxO
+    -> [(TxIn, TxOut)]
+utxoCovering txOuts utxo =
+    L.sortOn (coin . snd)
+    $ L.foldl (\acc o ->
+                   rotate 1
+                   $  L.dropWhile (\(_,_o) -> coin _o < coin o) acc
+              ) utxoSorted
+    $ outputsSorted txOuts
+    where
+        utxoSorted = L.sortOn (coin . snd)
+                     $ (Map.toList . getUTxO) utxo
+        rotate :: Int -> [a] -> [a]
+        rotate _ [] = []
+        rotate n xs = zipWith const (drop n (cycle xs)) xs
```

Comment:
I'd suggest a different approach here to avoid having too much logic in the test:

- change the precondition to something like `isRight . runExceptT . largestFirst ...`. In the end, that's really what the pre-condition is about here, checking that we can perform the coin selection! So, let's just perform it and assume that it's right. On small sets of value, it should be fast enough to not be too slow to evaluate with quickcheck.

Reply:
really do not know how to use `isRight` with `runExceptT` ...